File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:
    vector<string>v;
    void find(int &n,string s,int open,int close){
        if(s.length()==2*n){
            if(open==close){
                v.push_back(s);
            }
            return;
        }
        if(close<open){
            find(n,s+')',open,close+1);
        }
        find(n,s+'(',open+1,close);
    }
    vector<string> generateParenthesis(int n) {
        string s="";
        find(n,s,0,0);
        return v;
    }
};

File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    void find(int n, ArrayList<String>ans,int close,int open,String s){
        if(s.length()==n*2){
            if(open==close){
                ans.add(s);
            }
            return;
        }
        if(open>close){
            find(n,ans,close+1,open,s+')');
            find(n,ans,close,open+1,s+'(');
        }
        else{
            find(n,ans,close,open+1,s+'(');
        }
    }
    public List<String> generateParenthesis(int n) {
        ArrayList<String>arr=new ArrayList<>();
        find(n,arr,0,0,"");
        return arr;
    }
}